Orckestra.Widget.FilteredSelector/Orckestra.Widget.FilteredSelector/WidgetProvider/FilteredSelectorWidgetProcessingManager.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+
+using Composite.Core.Types;
+using Composite.Core.WebClient.Renderings.Page;
+using Composite.Data;
+
+namespace Orckestra.Widget.FilteredSelector.WidgetProvider
+{
+    internal static class FilteredSelectorWidgetProcessingManager
+    {
+        public static IEnumerable<Guid> GetAssociatedPageIds(Guid pageId, SitemapScope sitemapScope) => PageStructureInfo.GetAssociatedPageIds(pageId, sitemapScope);
+        private static IEnumerable<KeyValuePair> GetOptionsForDefault(Type t, Guid pageId, SitemapScope sitemapScope)
+        {
+            IEnumerable<Guid> availablePageIds = GetAssociatedPageIds(pageId, sitemapScope).Distinct();
+
+            using (DataConnection con = new DataConnection())
+            {
+                MethodInfo generic = typeof(DataConnection).GetMethod("Get").MakeGenericMethod(t);
+                IEnumerable<KeyValuePair> result = from a in (IEnumerable<IPageRelatedData>)generic.Invoke(con, null)
+                                                   join b in availablePageIds
+                                                   on a.PageId equals b
+                                                   select new KeyValuePair(a.GetKeyProperties().First(k => k.Name == "Id").GetValue(a).ToString(), a.GetLabel());
+                return result;
+            }
+        }
+
+        internal static IEnumerable GetParameters(string parameters)
+        {
+            Match match = Regex.Match(parameters, 
+                $"{Constants.pageIdParamName}:\"(.+?)\";{Constants.typeNameParamName}:\"(.+?)\";{Constants.sitemapScopeIdParamName}:\"([0-9]+)\"");
+
+            if (match.Success == false)
+            {
+                throw new ArgumentException(string.Format(Resources.default_text.FilteredSelectorWidgetProcManEx1, parameters));
+            }
+
+            Type type = TypeManager.GetType(match.Groups[2].Value);
+            Guid pageId = new Guid(match.Groups[1].Value.ToString());
+            SitemapScope sitemapScope = (SitemapScope)Enum.Parse(typeof(SitemapScope), match.Groups[3].Value.ToString());
+
+            foreach (KeyValuePair pageItem in GetOptionsForDefault(type, pageId, sitemapScope))

```

Comment:
If that's the only usage, you can make it return IEnumerable<KeyLabelPair>and not use "yield return", but just

`return GetOptionsForDefault(type, pageId, sitemapScope)`

Reply:
I propose to keep the original code. On C1 6.8 this change will cause an error.

Steps to reproduce:
1. To install the package.
2. To create a DataFolder A.
3. To create a DataFolder B with a referenced field to a DataFolder A and with using a FilteredSelector as a selector widget.
4. To add a DataFolder A and a DataFolder B to some page in Content perspective.
5. To try to add a new DataFolder B element. 

Error :

> System.InvalidOperationException: "Malformed Selection configuration. The Selected binding type 'Composite.Core.Types.KeyValuePair' does not have a property named 'Label' exception"

Reply:
You can use Composite.Plugins.Forms.WebChannel.UiControlFactories.KeyLabelPair instead of KeyValuePair. An alternative is to use GetOptionsForDefault(type, pageId, sitemapScope).Select(o => ....).

In general, it is better to returb IEnumerable instead of "yield return" syntax, as the latter has a side effect of producing a much less readable de-compiled code, when using tools like ILSpy.

Thanks for applying the fixes, please use the "Squash and merge option" when doing the merge, as there're a lot of commits :)

Reply:
> please use the "Squash and merge option"

Deal :) 

Updated